pkg/telemetry/client.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package telemetry collects the CLI metrics and sends the telemetry data to supercollider
+package telemetry
+
+import (
+	"encoding/json"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/vmware-tanzu/tanzu-cli/pkg/buildinfo"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/cli"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/common"
+	configlib "github.com/vmware-tanzu/tanzu-plugin-runtime/config"
+	configtypes "github.com/vmware-tanzu/tanzu-plugin-runtime/config/types"
+)
+
+var once sync.Once
+
+var client MetricsHandler
+
+type MetricsHandler interface {
+	UpdateCoreCommandMap(coreCMDs ...*cobra.Command)
+	AddToPluginsByTarget(target configtypes.Target, plugin *cli.PluginInfo)
+	UpdateCmdPreRunMetrics(cmd *cobra.Command, args []string) error
+	UpdateCmdPostRunMetrics(metrics *PostRunMetrics) error
+	SaveMetrics() error
+	SendMetrics() error
+}
+
+type telemetryClient struct {
+	currentOperationMetrics      *OperationMetricsPayload
+	coreCommandMap               map[string]bool
```

Comment:
Not sure if there are additional plans for coreCommandMap, but as of now it looks like its purpose is purely to capture a table of top-level core commands and rely of checking any Cmd's CommandPath() against it to determine if the command is a core command or not.

But I am thinking if is necessary at all. It's not so much the complexity of the changes since the implementation is straightforward, but I wonder if the telemetry client is making too many assumptions on how to interpret each command. I can imagine a scenario in the future when the second token in the command path is found in this map but the actual sub command could be implemented by a plugin.

Given that the Command tree is fully constructed by the time any Command in the tree is execute-d, have you considered finding a way to distinguish a core Command from a plugin-based Command by examining the Command object directly?  It might be possible because every plugin-based Command is constructed by GetCmdForPlugin(), so there might be some opportunity to explicity identify this category of Commands.

Reply:
Great point!! Thanks for brining this up. Updated the PR to use the command annotation from the Command directly and wouldn't need `coreCommandMap` anymore.